### ISSUE TITLE
Enable overload of `__set__` for descriptors used in dataclasses

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#5759]
If you have a question about a behavior that you’re seeing in pyright, consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Is your feature request related to a problem? Please describe.**
I want to modify the accepted type for a field depending on a generic argument given to the class.
Using descriptors who have overloads for `__set__` in a dataclass (or class having `@dataclass_transform` applied) results in an unexpected type hint for the generated `__init__`. I'd expect the same hint as using overloaded `__set__` in a normal class.

Sample code:
```python

from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...

    @overload
    def __set__(self, owner: "Foo[Literal[False]]", value: int) -> None:
        pass

    @overload
    def __set__(self, owner: "Foo[Literal[True]]", value: str) -> None:
        ...

    def __set__(self, owner, value): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False])

class Foo(Generic[TUseStr]):

    x: Desc
```

Assigning `Foo.x` requires the expected types:
```python

Foo[Literal[True]]().x = "hello world"
Foo[Literal[True]]().x = 5 # Cannot assign member "x" for type "Foo[Literal[True]]"
                     #  "Literal[5]" is incompatible with "Desc"
# kinda unexpected error message here though?
# I'd rather expect the error message you get when trying to actually assign a `Desc` instance

Foo[Literal[True]]().x = Desc() # Cannot assign member "x" for type "Foo[Literal[True]]"
                         #  No overloads for "__set__" match the provided arguments

# Same for Foo[Literal[False]] of course
Foo[Literal[False]]().x = 5
Foo[Literal[True]]().x = "hello world" # Cannot assign member "x" for type "Foo[Literal[False]]"
                                 #  "Literal['hello world']" is incompatible with "Desc"
```

Now - if we have `Foo` be a dataclass instead of a regular class the fact `__set__` is overloaded completely trips up pyright...

In the following code we get the error that `__set__` has no overloads - note having at most a single overload restricting on `Foo` or if `Foo` isn't a dataclass the error doesn't occur (I've no idea what's happening here tbh).

```python

from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...
    
    @overload
    def __set__(self, owner: object, value: str) -> None: # "__set__" is marked as overload, but additional overloads are missing
        ...
    
    @overload
    def __set__(self, owner: "Foo[Literal[True]]", value: str) -> None:
        ...

    @overload
    def __set__(self, owner: "Foo[Literal[False]]", value: int) -> None:
        ...

    def __set__(self, owner, value): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False])

@dataclass
class Foo(Generic[TUseStr]):

    x: Desc
```

There are quite a few additional oddities:

When using only a single overload with having the dataclass the overloads implementation is used for `__init__`.
```python


from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...
    
    @overload
    def __set__(self, owner: Any, value: str) -> None:
        ...

    @overload
    def __set__(self, owner: "Foo[Literal[False]]", value: int) -> None:
        ...


    def __set__(self, owner, value: str | int | bool): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False])


@dataclass
class Foo(Generic[TUseStr]):

    x: Desc

reveal_type(Foo[Literal[False]].__init__)
# Type of "Foo[Literal[False]].__init__" is "(self: Foo[Literal[False]], x: str | int | bool) -> None"
```

And having multiple overloads for `__set__` generally trips up pyrights `__init__` and it defaults to expecting an instance of the descriptor itself.

```python 

from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...
    
    @overload
    def __set__(self, owner: Any, value: str) -> None:
        ...

    @overload
    def __set__(self, owner: Any, value: int) -> None:
        ...


    def __set__(self, owner, value: str | int | bool): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False])


@dataclass
class Foo(Generic[TUseStr]):

    x: Desc

reveal_type(Foo[Literal[False]].__init__)
# Type of "Foo[Literal[False]].__init__" is "(self: Foo[Literal[False]], x: Desc) -> None"
```

Some very weird error message follows, but I think this is because pyright doesn't understand the descriptor in general

```python


from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...
    
    @overload
    def __set__(self, owner: Never, value: str) -> None: # "__set__" is marked as overload, but additional overloads are missing
        ...
    
    @overload
    def __set__(self, owner: "Foo[Literal[True]]", value: str) -> None:
        ...

    @overload
    def __set__(self, owner: "Foo[Literal[False]]", value: Any) -> None:
        ...



    def __set__(self, owner, value): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False])


@dataclass
class Foo(Generic[TUseStr]):

    x: Desc

Foo[Literal[False]](x=Any()).x = Any() # no error here
Foo[Literal[True]](x=Any()).x = Any() # Cannot assign member "x" for type "Foo[Literal[True]]"
                               #    "Any" is incompatible with "Any" 
```

**Describe the solution you’d like**
I'd expect that the type of `x` in the generated `__init__` is a union of all types of `value` where the overload is applicable considering the type of the instance that's currently being created.

To be more precise - I'd expect the following behaviours:
```python

from typing import *
from dataclasses import *

class Desc:

    def __get__(self, owner, type) -> Any: ...
    
    @overload
    def __set__(self, owner: "Foo[Literal[True]]", value: str) -> None:
        ...

    @overload
    def __set__(self, owner: "Foo[Literal[False]]", value: int) -> None:
        ...

    @overload
    def __set__(self, owner: "Foo[Any]", value: bool) -> None:
        ...


    def __set__(self, owner, value): ...


TUseStr = TypeVar("TUseStr", Literal[True], Literal[False], object)


@dataclass
class Foo(Generic[TUseStr]):

    x: Desc

reveal_type(Foo[Literal[True]].__init__)
# Type of "Foo[Literal[True]].__init__" is "(self: Foo[Literal[True]], x: int | bool) -> None"

reveal_type(Foo[Literal[False]].__init__)
# Type of "Foo[Literal[False]].__init__" is "(self: Foo[Literal[False]], x: str | bool) -> None"

reveal_type(Foo[object].__init__)
# Type of "Foo[object].__init__" is "(self: Foo[object], x: bool) -> None"
```

I was considering to submit a bug report but to me this seems to be out of scope for that.

**Comment by erictraut** [comment]
I applaud you for your creativity, but I don't think this is going to work. You're attempting to leverage a number of aspects of the Python type system that have complex edge cases and undefined behaviors. Such behaviors tend to not compose well. For example, the use of descriptors in dataclasses were never contemplated by the authors of PEP 557 (by their own admission).

As you can probably guess, dataclasses require a bunch of special-case logic in a type checker. One of the complexities is that an `__init__` method must be synthesized, and the types of the parameters for this `__init__` method depend on the types of the fields defined in the class. In the typical case, these types can be evaluated directly from the type annotations within the dataclass class definition. However, in the case of a descriptor, the type must be determined by evaluating the `__set__` method of the descriptor and extracting the type from the `value` parameter. In your case, the `__set__` method is a complex overload. There are two problems here. First, some of the overloaded signatures of `__set__` circularly reference `Foo`. This causes a bunch of the odd behaviors you're seeing because pyright isn't able to handle the circular reference. Second, the overload makes the `value` parameter type (and therefore the type associated with the `x` parameter in the synthesized `Foo.__init__` method) ambiguous because the type depends on the specialized type of `Foo`, but the `__init__` method is common to all specializations of the dataclass.

It would be theoretically possible to support your scenario, but it would require adding massive complexity to code that is already very complex. Since this is such an odd usage — one that I doubt any other pyright user will ever consider, I don't think it makes sense to add such support.

Pyright already goes well beyond what mypy does to handle some of these features (such as descriptors in dataclasses).

As a workaround, you could specify `init=False` in the dataclass and define your own `__init__` method rather than relying on pyright (and the runtime) to synthesize one for you.
